### ISSUE TITLE
BC6H/BC7 support

### ISSUE THREAD

**Issue by Krakean** [vpenades/TextureSquish#2]
Hello
Can you please implement support for these two formats?
Seems like original libsquish do not implemented them, due to fact that specifications are widely available:
1. https://www.khronos.org/registry/OpenGL/extensions/ARB/ARB_texture_compression_bptc.txt
2. https://msdn.microsoft.com/en-us/library/hh308953.aspx
3. https://msdn.microsoft.com/en-us/library/hh308952.aspx

**Comment by vpenades** [comment]
I have it in my roadmap, along refactoring the other encoders to try improve the compression performance; but I don't know when I will have time to write an implementation.

**Comment by Krakean** [comment]
@vpenades 
>I have it in my roadmap

Good to hear. Looking forward for this feature! :) 
Then I would like to use your library in my game world editor to produce dxt/bc dds/ktx compressed textures from their raws like tga/psd counterparts.

>along refactoring the other encoders to try improve the compression performance; 

Current encoders performance is more or less well enough, in my opinion. 
Better to concentrate on BC6H/BC7, I suppose :)

**Comment by vpenades** [comment]
@Krakean Actually, it's the opposite; we were planning to use this library in monogame for texture compression, and it has been pulled back just because performance issues, so for now, improving performance has priority over supporting BC6H/BC7.

For example, on large projects, it was reported to me that build times went from 5 minutes with NVTT to 40 minutes with TextureSquish, which is unaceptable. As a consequence, monogame has reverted back to using NVTT.

If you want a short term solution, I suggest you to look at [NVTT fork at monogame](https://github.com/MonoGame/nvidia-texture-tools) , which fixed some issues related to interacting with C#

In the long run everybody wants a pure NET solution, but it needs to have good enough performance to be worth replacing NVTT with it, and that will take more time than a simple solution.

**Comment by Krakean** [comment]
@vpenades 
Got it :)

>we were planning to use this library in monogame for texture compression

Hopefully it still will be engine-agnostic solution, because in my case I want to use your library within my in-house engine

> it was reported to me that build times went from 5 minutes with NVTT to 40 minutes with TextureSquish, which is unaceptable. 

May be some Span<T> & unsafe & System.Memory could solve the problem?
By the way, do you test it on .NET Core 2.0 in context of performance?

**Comment by vpenades** [comment]
@Krakean 
> May be some Span & unsafe & System.Memory could solve the problem?

yes, but not only that, the inner loops need to be refactored, because they were copied straight away from the c++ squishlib.

You can follow the whole discussion here: https://github.com/MonoGame/MonoGame/pull/6102

> By the way, do you test it on .NET Core 2.0 in context of performance?

Not yet, but given it's pure net standard code, I would not expect many differences. What I haven't tried is compiling to native with roslyn.

**Comment by Krakean** [comment]
@vpenades 
By the way, just curious, have you took a look on UpkManager.Dds (https://github.com/stricq/UPKManager) ? 
Seems like they also converted Libsqsuish to C#, but may be their compressor is faster (at least, if reading commits, they parallelized compression code) and this code can help you somehow? Check it out, when you will have time. UpkManager.dds is standalone, so you don't need rest of modules to compile & test it.

**Comment by vpenades** [comment]
@Krakean I took a look at the code at https://github.com/stricq/UPKManager, as mine, it's pretty much a straight forward reimplementation of libsquish. They use unsafe code to boost performance a bit... but you'll get a 10% performance increase at best. The main problem is that the code has a lot of branching, which is what's making the code so slow. In order to get an order of magnitude performance improvement, the whole thing must be redesigned.

TextureSquish already uses parallel compression, and SIMD vectors for faster maths.. but still not fast enough...
